src/Formatters/AbstractCodeFormatter.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.Tools.Formatters
+{
+    internal abstract class AbstractCodeFormatter : ICodeFormatter
+    {
+        public async Task<Solution> FormatAsync(
+            ILogger logger, 
+            Solution solution, 
+            ImmutableArray<(Document, OptionSet)> formatableDocuments,
+            CancellationToken cancellationToken)
+        {
+            var formattedDocuments = FormatFiles(logger, formatableDocuments, cancellationToken);
+            return await ApplyFileChangesAsync(logger, solution, formattedDocuments);
+        }
+
+        protected abstract ImmutableArray<(Document, Task<SourceText>)> FormatFiles(
+            ILogger logger, 
+            ImmutableArray<(Document, OptionSet)> formatableDocuments, 
+            CancellationToken cancellationToken);
+
+        private static async Task<Solution> ApplyFileChangesAsync(
```

Comment:
I assume the purpose of this method will be made clear in the fullness of time

Reply:
I'll add comments. =)